### PR TITLE
[WebXR] Use ANGLE extension to signal compositor completion event

### PR DESCRIPTION
**Pull request by djg**
#### 2a4dc7e13fde3c6dd2995ea7702cf47ea28269de
<pre>
[WebXR] Use ANGLE extension to signal compositor completion event
<a href="https://bugs.webkit.org/show_bug.cgi?id=246589">https://bugs.webkit.org/show_bug.cgi?id=246589</a>
rdar://problem/101212895

Reviewed by Dean Jackson.

Use MTLSharedEvent, provided by WebXR system compositor, that signals when scene
rendering is completed using the EGL_ANGLE_metal_shared_event_sync extension.

To enable, set USE_MTLSHAREDEVENT_FOR_XR_FRAME_COMPLETION to 1

* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::endFrame):
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
* Source/WebCore/PAL/pal/spi/cocoa/MetalSPI.h:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.mm: Renamed from Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.cpp.
(WebCore::newSharedEventWithMachPort):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::platformSupportsMetal):
(WebCore::initializeEGLDisplay):
(WebCore::GraphicsContextGLCocoa::newSharedEventWithMachPort):
(WebCore::GraphicsContextGLCocoa::createSyncWithSharedEvent):
(WebCore::GraphicsContextGLCocoa::destroySync):
(WebCore::GraphicsContextGLCocoa::clientWaitSyncWithFlush):
* Source/WebCore/platform/xr/PlatformXR.h:
(PlatformXR::Device::FrameData::LayerData::encode const):
(PlatformXR::Device::FrameData::LayerData::decode):

Canonical link: <a href="https://commits.webkit.org/255613@main">https://commits.webkit.org/255613@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/6a331e94e018d20f9f2ad40c8daf3f9347b6352c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93077 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2278 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23712 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102785 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163037 "Found 30 new test failures: accessibility/loading-iframe-updates-axtree.html, css3/filters/effect-blur.html, css3/filters/effect-combined.html, css3/filters/effect-contrast.html, css3/filters/effect-grayscale.html, css3/filters/effect-hue-rotate.html, css3/filters/effect-invert.html, css3/filters/effect-opacity.html, css3/filters/effect-saturate.html, css3/filters/effect-sepia.html ... (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97079 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2286 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30605 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85463 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98902 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98742 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1574 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79551 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28492 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83488 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/83243 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71601 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/84398 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36999 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17135 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/79458 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34811 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18361 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27509 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/3888 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38682 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40911 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/82091 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40611 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37564 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18570 "Passed tests") |

### CODE REVIEW

File: Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp [hunk 1]
```diff
@@ -228,6 +233,17 @@ void WebXROpaqueFramebuffer::endFrame()
...
+    if (m_completionEvent) {
+        auto gCGL = static_cast<GraphicsContextGLCocoa*>(&gl);
+        auto completionSync = gCGL->createSyncWithSharedEvent(m_completionEvent, m_renderingFrameIndex);
+        ASSERT(completionSync);
+        constexpr uint64_t kTimeout = 1'000'000'000; // 1 second
+        gCGL->clientWaitSyncWithFlush(completionSync, kTimeout);
+        gCGL->destroySync(completionSync);
+    } else
+#endif
+
```
Comment (grorg) [inline Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp]: I think I'd prefer to put the `gl.finish()` part in an `#else` block too.

**Comment by grorg** [comment]
Compilation is failing because non-Internal builds don't know what `MTLSharedEventHandle` and `MTLSharedTextureHandle` are. I'm confused by the latter, since you didn't add that.

**Comment by grorg** [comment]
Maybe you need to `#include <Metal/MTLEvent.h>` ?

**Review by grorg — APPROVED** [review]
Approving once it compiles.

**Comment by djg** [comment]
> Maybe you need to `#include <Metal/MTLEvent.h>` ?

I'll try that. The thing is it all compiles on my machine.

**Pull request closed without merging**